[Rules] Optimize resource usage in parseTemplate function

TD-er opened the pull request:
The `parseTemplate` function did make a lot of memory re-allocations.
This is now reduced to a minimum.

Also added some build optimizations to (hopefully) reduce build times.

TD-er committed:
[Rules] Reduce memory usage when parsing rules templates

The parsetemplate function was almost always shown as function with the lowest free memory when called.
The function did make a lot of memory re-allocations and was very hard to read.

This commit is about reducing the memory allocations and speed up the parsing.
Will also fix [this issue reported on the forum](https://www.letscontrolit.com/forum/viewtopic.php?p=39219#p39219)

TD-er committed:
[Modbus RTU] Add "timeout" and "no data" as error

Does not yet act on the new errors, but at least it should be possible in the plugins to detect whether any data was received to see if a device is present.

See also https://github.com/letscontrolit/ESPEasy/issues/2616

TD-er committed:
[Build] No longer ignore IR libraries to allow IR builds from Vagrant

The builds not using the IR libraries will not increase in size by no longer ignoring these libraries.

TD-er committed:
[Travis] Do not remove old builds to reduce build time

The clean step may no longer be needed, which makes it more useful to have the last built directories restored from the cache and hopefully reduce build times.

TD-er committed:
[Travis] Add .pio dir as cache directory

TD-er commented:
A [test build - ESPEasy_mega-20190910-6-PR_2617.zip](https://www.dropbox.com/s/68u55dhvayhokml/ESPEasy_mega-20190910-6-PR_2617.zip?dl=0)

Please let me know if something of the following is broken with this build:
- Parsing references to task values (e.g. `[bme#temp]` )
- SenseAir plugin

TD-er committed:
[Build] Remove some build types from the nightly build

Mainly core 2.4.2 and double core 2.5.2 builds

TD-er committed:
[Travis] Change Ubuntu 14.04 to 18.04 for builds

We were still using "Trusty", changed to "Bionic" to get newer versions of GCC etc.

TD-er committed:
[Rules] Allow transformations on [var#n] and add [int#n]

Also add some caches to find task-value names.
Fix some minor issue when referring [var#n] variables. These were always rounded to 2 decimals.
Also added [int#n] to address the same variables as with [var#n] but rounded to integer values, to be used in rules for more reliable comparing.

TD-er committed:
[Cache] Add cache to help find task value name.

TD-er committed:
[PIO] Update core 2.5.2 to espressif8266@2.2.3

uzi18 commented on src/ESPEasyRules.ino:
```diff
@@ -156,42 +156,83 @@ String rulesProcessingFile(const String& fileName, String& event) {
-        line.replace("\r", "");
+      switch (static_cast<char>(data))
+      {
+        case 10: // "\n"
```
just case '\n'

TD-er replied:
True, that was a left-over from when the compiler said "no" this morning :)

uzi18 replied:
same for 32  - ' '

TD-er replied:
See https://github.com/letscontrolit/ESPEasy/pull/2617/commits/b44abc01936e815db8e4c90e099a6315ac51465e

uzi18 commented on src/ESPEasyRules.ino:
```diff
@@ -156,42 +156,83 @@ String rulesProcessingFile(const String& fileName, String& event) {
+  byte ifBlock     = 0;
+  byte fakeIfBlock = 0;
+
+  std::vector<byte> buf;
```
std::vector<byte> buf is better than byte array?

TD-er replied:
The SPIFFS_CHECK can return the function, so there is no delete of the buf.
std::vector (like any other STL container) does clean up the allocated stuff when destructed.

uzi18 commented on src/ESPEasyRules.ino:
```diff
@@ -156,42 +156,83 @@ String rulesProcessingFile(const String& fileName, String& event) {
+          commentFound      = false;
+          break;
+        }
+        case 13:   // "\r", Just skip this character
```
case '\r'

uzi18 commented on src/ESPEasyRules.ino:
```diff
@@ -202,6 +243,25 @@ String rulesProcessingFile(const String& fileName, String& event) {
   return "";
 }
 
+void replace_EventValueN_Argv(String& line, const String& argString, unsigned int argc)
```
maybe count argc from 0 to 4 and for 0 don't do "eventvalue += argc;" but only assign "argc =1;" so you can drop lines from 257-260

TD-er replied:
Hmm that's a good one.
Will do :)

TD-er committed:
[Rules] Some minor code improvements (suggests from @uzi18 )

uzi18 commented on src/ESPEasyRules.ino:
```diff
@@ -248,16 +248,15 @@ void replace_EventValueN_Argv(String& line, const String& argString, unsigned in
+
+  if (argc != 0) {
+    eventvalue += argc;
+  }
```
else argc =1;

TD-er replied:
Nope, argc is given as function parameter and I extended the for loop calling the function.

TD-er replied:
Ah, I see what you mean.....

TD-er commented:
@uzi18 Have you also tested the PR?
In my tests, the references to variables still seem to work and the added caches and code optimizations did reduce the timings of the OLED Framed plugin quite a bit. (137 msec on average to < 80 msec)

TD-er committed:
[C014 Homie] Use cached functions to find task value by name

TD-er committed:
[Rules] Compatibility Replace %eventvalue% by using the 1st argument

As noted by @uzi18

TD-er committed:
[Build] Remove debug logs from 1M builds

TD-er committed:
[Rules] Only match task value names which can be set

Otherwise there could be a match on variables which are not active or usable.

uzi18 commented:
@TD-er no have almost no time for coding/testing right now.
about std:vector look here: https://github.com/SmingHub/Sming/tree/develop/Sming/Wiring
you can find there WVector,WHashMap,SplitString,FIFO,FILO etc. classes/templates maybe you will find something interesting :)

TD-er committed:
Only include HttpClient when Settings Archive is being used

TD-er committed:
[cleanup] make it compile again on ArduinoIDE

Changes to file locations to make it compile again under Arduino IDE.
Basically `#include ""` and file locations have been fixed.

TD-er committed:
[Custom.h] Storage layout parameters moved and wrapped in #ifndef

TD-er committed:
[Storage layout] Add suggestions for 24 tasks layout

TD-er committed:
[Custom.h] Allow for other defines to be set by the user.

TD-er committed:
[Build] Fix include path to make ArduinoIDE able to build again

TD-er committed:
[Travis CI] Reduce Travis test build duration

Specify only a subset of PIO envs to build, to make sure it will build within the 50 minutes time limit, even when the caches are cleared by PlatformIO.

TD-er committed:
[PIO] Use original platform for core 2.6.0, not Tasmota's one

Was using https://github.com/Jason2866/platform-espressif8266.git#Tasmota
This may lead to issues when running `pio update`

uzi18 reviewed (commented):
why you revert to original?

TD-er committed:
[Travis] Disable Travis cache

TD-er commented:
> why you revert to original?

See https://github.com/platformio/platformio-core/issues/1223#issuecomment-532623945

But I will switch back now, right after clearing all Travis caches.

TD-er committed:
[Build] Make sure Custom.h is leading during compilation.

Also allowing a new define: USE_NON_STANDARD_24_TASKS

TD-er committed:
[ESP32] Fix build when SettingsArchive is set in Custom.h

TD-er committed:
Fix using includes to get defines right during compile

For detailed explanation, see https://github.com/letscontrolit/ESPEasy/issues/2621#issuecomment-533673956

TD-er committed:
Move all header files in Commands/*.h to src/Commands/ for ArduinoIDE

Arduino IDE needs include files in a subdirectory called `src` or some directory in `src`

TD-er committed:
[Build] Fix path change in include

TD-er committed:
Initial step for proper declaring global variables in .h files

TD-er committed:
[Build] Another forgotten path update in moved .h file

TD-er committed:
Move defines out of ESPEasy-Globals.h

TD-er committed:
Move Plugin administration from ESPEasy-Globals.h to Globals/Plugin.h

TD-er committed:
Move Logging from ESPEasy-Globals.h to Globals/Logging.h

TD-er committed:
Move Cache from ESPEasy-Globals.h to Globals/Cache.h

TD-er committed:
Move RTC from ESPEasy-Globals.h to Globals/RTC.h

TD-er committed:
Move Nodes from ESPEasy-Globals.h to Globals/Nodes.h

TD-er committed:
Move FactDefPrefs & CRCValues from ESPEasy-Globals.h to Globals/...h

TD-er committed:
Move Device array from ESPEasy-Globals.h to Globals/Device.h

TD-er committed:
Move SecuritySettings from ESPEasy-Globals.h to Globals/SecuritySettings.h

TD-er committed:
Move ExtraTaskSettings from ESPEasy-Globals.h to Globals/...h

TD-er committed:
Move Settings from ESPEasy-Globals.h to Globals/Settings.h

TD-er commented:
Would be great if this one is tested so far 
I have not yet fixed the esp32 with 24 tasks.

TD-er committed:
Split Diagnostic.h & ESPEasyWiFiEvent.h into .h/.cpp

TD-er committed:
[Arduino IDE] Make include paths relative to file location

TD-er committed:
[Build] Forgotten include of MQTT related global variable.

TD-er committed:
[Cleanup] Remove FIXME/TODO remarks which are now done.

TD-er committed:
[Arduino IDE] Fix path in includes.

TD-er committed:
Uncrustify P022_PCA9685

TD-er committed:
[PCA9685] Add range checks and memory optimization

TD-er committed:
[ESP32] Fix storage layout on ESP32 settings file.

TD-er committed:
[PIO] Switch to core_stage for core 2.6.0 builds

To be able to use this merged PR: https://github.com/esp8266/Arduino/pull/6484

TD-er committed:
[Build] Fix pre_extra_script.py  build Cpp defines

To append a list to an existing list in Python, you must used extend.
When appending a list, it will create list at the end of the existing list.

TD-er committed:
[Build] Add Custom ESP32 and disable plugins to reduce binary size

Disabled 
- Settings Archive
- NeoClock
- Candle

TD-er committed:
Define build info from ESPEasy_common.h

TD-er committed:
[Docs] Describe project directory structure

TD-er committed:
Turn SettingsStruct into templated variable

This will be useful later when making the number of tasks dynamic, or converting settings into another settings file using different amount of tasks.
